### DISCUSSION TITLE
bind_tweets error

### DISCUSSION THREAD

**Discussion by keoghca** [cjbarrie/academictwitteR#197] [Q&A] [answered]
I have just collected data using academictwitteR and am trying to use the bind_tweets function to convert the JSON files to an R data frame. Here's the code I used: 

       tweets <- bind_tweets(data_path = "alltweetdata/", output_format = "tidy")

Here's the error message: 

       Error: cannot allocate vector of size 13.1 Mb
       In addition: Warning message:
       closing unused connection 3 (alltweetdata/data_1409120214051139589.json) 

Any suggestions?

**Comment by cjbarrie** [comment]
How large is the data? It looks like an error due to insufficient local memory

**Comment by keoghca** [comment, marked as answer]
I'm not sure exactly how large it is, but I imagine it's a few million tweets. Would getting a solid state help?
________________________________

**Reply by chainsawriot** [reply]
Before getting anything, please check the size (in the megabytes sense) of your `alltweetdata` directory.

Please be aware that `bind_tweets` puts all these data into your computer's main memory for processing. It is difficult to estimate the amount of memory needed. But a rough rule of thumb is at least 3x; i.e. if your `alltweetdata` is 10G, you need at least 30G of RAM to put all of them in the main memory. The manipulation tasks for converting it to tidy data frame requires even more memory. Probably 10x.

Having said that, this procedure can be easily solved by `divide et impera`. The problem is, you don't need to put them all in memory. But it is not a feature provided by this package. You need to develop it yourself. MongoDB, for example, supports JSON. You can import those JSON files collected by the package and put them **one-by-one** into MongoDB.

**Reply by keoghca** [reply]
The alltweetdata is 4.48G (7,069 files) and my computer only has 8G of RAM. If I understand correctly, using an external hard drive would not help since the bind_tweets function brings the data into the computer's main memory anyway. Is that right? If so, would MongoDB be any different? Also, since the files would need to be added one-by-one into MongoDB, would I be able to then use bind_tweets? 

(Apologies is these are dumb questions. I am not a developer--I'm an academic working on my first project of this sort that is apparently a bit over my head.)

**Reply by chainsawriot** [reply]
As you may know, I have opened a new issue on reworking `bind_tweets`. I am not a developer too, but still I need to develop.

**Reply by keoghca** [reply]
I just checked out the new issue. A convert_json function would be great. Thanks so much for your help!

**Reply by chainsawriot** [reply]
`convert_json` is now available. It solves a small part of your problem.

**Reply by keoghca** [reply]
Just tested it out and it worked! Thank you so much!!

**Reply by chainsawriot** [reply]
@keoghca You may consider also using `academicquacker` from [here](https://github.com/chainsawriot/academicquacker).

**Reply by keoghca** [reply]
@chainsawriot Thanks! I just tried it, but am getting an error...

con <- quack(dir, db = "D:/Tweets/alltweetdata.duckdb", db_close = TRUE)

Results in this error: 

Error in file.path(data_path) : 
  cannot coerce type 'closure' to vector of type 'character'

**Reply by chainsawriot** [reply]
You should do this:

```r
con <- quack("alltweetdata", db = "D:/Tweets/alltweetdata.duckdb", db_close = TRUE)
```

**Reply by keoghca** [reply]
Got it now--very cool! Thanks so much!!

**Comment by shmuhammadd** [comment]
@chainsawriot  I have an error below after running` convert_json`: 

```
Error in lapply(columnlist, simplify, simplifyVector = TRUE, simplifyDataFrame = TRUE,  :  could not find function "FUN"
```


what I did :

```

files <- list.files(here::here("tweets"), pattern = '\\.json$')

files_path <- fs::path("tweets", files )


df_json <- convert_json(files_path, output_format = "tidy")


```

However, `convert_json` works on single JSON file I tried, but it failed on this example because I have about 2,000 JOSN file.

**Reply by chainsawriot** [reply]
Read `?convert_json`, it says:

```
data_file: string, a single file path to a JSON file; or a vector of
          file paths to JSON files of stored tweets data saved as
          data__id_.json
```
So, it works only with `data_????.json` files, not all json files as your `list.files` yielded.

Do this instead:

```
data_path <- list.files(here::here("tweets"), full.names = TRUE, pattern = "data.+\\.json")
```

**Reply by shmuhammadd** [reply]
@chainsawriot it works. Thanks a lot.
